Allows empty borg shells to be disassembled.

zxaber opened the pull request:
:cl: Zxaber
tweak: Empty cyborg shells can now be disassembled with a wrench.
tweak: Using a screwdriver on an empty shell will now remove the cell, or swap it if you're holding a cell in your other hand.
/:cl:

Sometimes you end up with a borg shell sitting around from roundstart, and a better battery next to it waiting. It would be nice to replace the battery _before_ having an MMI to put in.

Besides, a functional borg can be disassembled, so it's weird than the empty shell cannot.

Flashes remain unbroken because the shell hasn't been used yet.

~~Adding the [Ready] tag, as all requested changes have been done.~~

zxaber committed:
Adds the ability to deconstruct empty cyborg shells.

zxaber committed:
Update robot_parts.dm

Repukan commented:
Not sure if it's in this codebase, but you should look into making it so using a screwdriver and a battery in that order lets you change the battery too.

Hell that'd be miles better than having to put everything back together just to change the battery.

zxaber committed:
Able to directly swap cells

Added ability to change out power cells with a screwdriver.

zxaber commented:
In the interest of not creating yet another state for the shell to be in, it will now directly swap power cells if you use the screwdriver on the cell with the power cell in your other hand.

@Repukan thanks for the idea.

zxaber committed:
Update robot_parts.dm

ninjanomnom commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -78,6 +78,52 @@
+					src.l_leg = null
+				if(r_leg)
+					src.r_leg.forceMove(T)
+					src.r_leg = null
```
You don't need all these `src.`

You can in fact search and replace that in your additions

zxaber replied:
Yep, looks like it runs fine without `src`. Why is that even there? Oh well, updated

ninjanomnom commented:
This is some ancient code you got here.

...I hate even looking at it.

zxaber committed:
Removing unnecessary `src` prefixes from code.

zxaber committed:
Update robot_parts.dm

zxaber committed:
Moved to wrench_act and screwdriver_act overrides

I think I did this right

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,54 @@
+		if(!user.transferItemToLoc(user.held_items[user.active_hand_index == 1 ? 2 : 1], src.chest))
+			chest.cell = null
+			return
+		to_chat(user, "<span class='notice'>You replace the power cell.</span>")
```
```
/obj/item/robot_suit/screwdriver_act(mob/living/user, obj/item/I) //Swaps the power cell if you're holding a new one in your other hand.
	var/obj/item/stock_parts/cell/temp_cell = user.is_holding_item_of_type(/obj/item/stock_parts/cell)
	if(!temp_cell)
		return
	if(!user.transferItemToLoc(temp_cell, chest))
		to_chat(user, "<span class='warning'>[temp_cell] is stuck to your hand, you can't put it in [src]!</span>")
		return
	if(!QDELETED(chest.cell))
		if(!user.put_in_hands(chest.cell))
			chest.cell.forceMove(drop_location())
	chest.cell = temp_cell
	to_chat(user, "<span class='notice'>You replace [src]'s power cell.</span>")
	return TRUE
```
this will put the old cell in the user's hands instead.

zxaber replied:
the line `to_chat(user, "<span class='notice'>There doesn't seem to unscrew from [src].</span>")` is for if no cell is found inside the shell, right?

zxaber committed:
Replaced screwdriver code

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,59 @@
+	if(QDELETED(temp_cell))
+		return
+	if(QDELETED(chest.cell))
+		to_chat(user, "<span class='notice'>There doesn't seem to unscrew from [src].</span>")
```
whoops. anything to unscrew*

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,59 @@
+	var/obj/item/stock_parts/cell/temp_cell = user.is_holding_item_of_type(/obj/item/stock_parts/cell)
+	if(QDELETED(temp_cell))
+		return
+	if(QDELETED(chest.cell))
```
`if(QDELETED(chest) || QDELETED(chest.cell))`  forgot a sanity check.

Cyberboss replied:
same here

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,59 @@
+		chest.cell.forceMove(drop_location())
+	chest.cell = temp_cell
+	to_chat(user, "<span class='notice'>You replace [src]'s power cell.</span>")
+	return TRUE
```
Actually, rethought this a bit. Why not just make it so you can always remove the powercell from the shell?

```
/obj/item/robot_suit/proc/put_in_hand_or_drop(mob/living/user, obj/item/I) //normal put_in_hands() drops the item ontop of the player, this drops it at the suit's loc
	if(!user.put_in_hands(I))
		I.forceMove(drop_location())
		return FALSE
	return TRUE

/obj/item/robot_suit/screwdriver_act(mob/living/user, obj/item/I) //Removes the current power cell, or swaps it if you're holding a new one.
	if(QDELETED(chest) || QDELETED(chest.cell))
		to_chat(user, "<span class='notice'>There doesn't seem to be anything to unscrew from [src].</span>")
		return

	var/obj/item/stock_parts/cell/temp_cell = user.is_holding_item_of_type(/obj/item/stock_parts/cell)
	var/swap_failed
	if(QDELETED(temp_cell)) //if we're not holding a cell
		swap_failed = TRUE
	else if(!user.transferItemToLoc(temp_cell, chest))
		swap_failed = TRUE
		to_chat(user, "<span class='warning'>[temp_cell] is stuck to your hand, you can't put it in [src]!</span>")

	put_in_hand_or_drop(user, chest.cell)

	if(swap_failed)
		to_chat(user, "<span class='notice'>You remove [chest.cell] from [src].</span>")
		chest.cell = null
		return

	chest.cell = temp_cell
	to_chat(user, "<span class='notice'>You replace [src]'s power cell.</span>")
	return TRUE
```

zxaber replied:
Presumably, the reason you can't make a cell-less shell in the first place is because it's nice as the borg player to _know_ that you won't spawn in with power problems. Sure, the robo player might use a depleted cell, but the borg can fix that themselves with a charging station, rather than having to sit there and ping or buzz until someone bothers to figure out what they need.

If you specifically need to use the cell somewhere quickly, the disassemble-via-wrench will be available, same as if you suddenly need to give someone a prosthetic limb.

ninjanomnom replied:
Just let the ghost know how much power there is then if you're worried about that

ExcessiveUseOfCobblestone replied:
Ghosts don't go in borgs, they go in MMIs that go into borgs. The ghost doesn't have an option in this situation outside of not opting into borgs at all.

I can assure you that issue won't happen anymore than roboticists immediately flashing the borg on spawn to remove their powercell.

zxaber replied:
Seems I'm outvoted. I'll add the code in a bit, after I test and reorganize it a bit.

Would anyone be opposed to me preventing an MMI to be inserted without a cell in the shell?

Cyberboss commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,59 @@
+					chest.cell = null
+				chest.forceMove(T)
+				new /obj/item/stack/cable_coil(T, 1)
+				chest.wired = 0
```
FALSE

zxaber replied:
What, specifically, is wrong with the code here? Aside from the interlaced `if` statements, it's the code used to deconstruct a functional borg.

ninjanomnom replied:
the 0 needs to be `FALSE`

zxaber replied:
Zero has been replaced with `FALSE`.

Cyberboss commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,59 @@
+
+/obj/item/robot_suit/screwdriver_act(mob/living/user, obj/item/I) //Swaps the power cell if you're holding a new one in your other hand.
+	var/obj/item/stock_parts/cell/temp_cell = user.is_holding_item_of_type(/obj/item/stock_parts/cell)
+	if(QDELETED(temp_cell))
```
just `!temp_cell`. is_holding_item_of_type should never return something being deleted

zxaber committed:
Swapping out qdeleted for an if-not

zxaber committed:
Adds a check for cell-less shells when MMI is added.

zxaber committed:
Update robot_parts.dm

zxaber committed:
Adds the ability to just remove the cell

Re-organized the screwdriver section as well, and added the ability to add a cell to a shell with no cell.

zxaber committed:
Whoops

Copypaste error

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,81 @@
+				put_in_hand_or_drop(user, chest.cell)
+				to_chat(user, "<span class='notice'>You replace [src]'s [chest.cell.name] with [temp_cell].</span>")
+				chest.cell = temp_cell
+				return TRUE
```
This can all be condensed down to 
```
	if(QDELETED(chest)) //can't remove a cell if there's no chest to remove it from.
		to_chat(user, "<span class='notice'>There isn't anything to remove from [src]!</span>")
		return

	var/obj/item/stock_parts/cell/temp_cell = user.is_holding_item_of_type(/obj/item/stock_parts/cell)
	var/swap_failed
	if(!temp_cell) //if we're not holding a cell
		swap_failed = TRUE
	else if(!user.transferItemToLoc(temp_cell, chest))
		swap_failed = TRUE
		to_chat(user, "<span class='warning'>[temp_cell] is stuck to your hand, you can't put it in [src]!</span>")

	if(chest.cell) //drop the chest's current cell no matter what.
		put_in_hand_or_drop(user, chest.cell)

	if(swap_failed) //we didn't transfer any new items.
		if(chest.cell) //old cell ejected, nothing inserted.
			to_chat(user, "<span class='notice'>You remove [chest.cell] from [src].</span>")
			chest.cell = null
		return

	to_chat(user, "<span class='notice'>You [chest.cell ? "replace [src]'s [chest.cell.name] with [temp_cell]" : "insert [temp_cell] into [src]"].</span>")
	chest.cell = temp_cell
	return TRUE
```

Same functionality and far easier to read imo.

zxaber replied:
When I went to add the battery insert part, I had re-formatted your code to avoid using extra variables unnecessarily. In hindsight, the nested if statements are probably less preferred. Replacing the screwdriver section.

ShizCalev commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,81 @@
+			chest.cell = user.held_items[user.active_hand_index == 1 ? 2 : 1]
+			if(!user.transferItemToLoc(user.held_items[user.active_hand_index == 1 ? 2 : 1], src.chest))
+				chest.cell = null
+				to_chat(user, "<span class='warning'>[user.held_items[user.active_hand_index == 1 ? 2 : 1]] is stuck to your hand, you can't put it in [src]!</span>")
```
var

optimumtact commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -62,6 +62,81 @@
 
+/obj/item/robot_suit/wrench_act(mob/living/user, obj/item/I) //Deconstucts empty borg shell. Flashes remain unbroken because they haven't been used yet
+	var/turf/T = get_turf(src)
+	forceMove(T)
```
why are you moving the shell to the turf as well?

zxaber replied:
Originally, I wanted to make sure all the parts ended up at the same spot, so it would pull the naked shell out of the player's hand (if they were holding it) and drop it on the ground. But there's probably no real reason to force this, so I'll remove that line.

zxaber committed:
Changed the screwdriver section from nested if statements to use a function

Also removed the line that forced the naked endoskeleton out of your hand in the wrench section.
